onnxruntime/test/common/tensor_op_test_utils.h

```diff
@@ -51,6 +51,58 @@ class RandomValueGenerator {
     return random_seed_;
   }
 
+  template <typename TValue>
+  std::vector<TValue>
+  Discrete(gsl::span<const int64_t> dims, const std::vector<TValue>& value_candidates) {
+    std::vector<TValue> values(detail::SizeFromDims(dims));
+    std::uniform_int_distribution<size_t> distribution(0, value_candidates.size() - 1);
+    // Tier 2 RNG. Use it if `Uniform` method causes large numerical errors
+    // (e.g., when elementwise relative error > 1e-3).
+    //
+    // The generated tensor is more numerically stable than other `Uniform` methods.
+    // If a test constantly fails, it's better to use `Discrete` method. For example,
+    // we call `Discrete` method to generate a tensor with shape [2, 3] and value candidates
+    // [-1, 0, 1] to eliminate the error propagation caused by * and / in matrix multiplication.
+    // To trigger mild error propagation in * and /, try value candidates [-1.5, -1, 0, 1, 1.5].
+    //
+    // Suggested value_candidates to alleviate numerical error (listed
+    // from smallest error to largest error):
+    //  [0]
+    //  [1]
+    //  [0, 1]
+    //  [-1, 0, 1]
+    //  [-2, 1, 0, 1, 2]
+    //  [-1.5, -1, 0, 1, 1.5]
+    //  [-2, -1.5, -1, 0, 1, 1.5, -2]
+    for (size_t i = 0; i < values.size(); ++i) {
+      auto index = distribution(generator_);
```

Comment:
I think this function should use a fixed `generator_`. If this function is called, it means caller wants ultimate stability and the current `generator_` may be changed by system random seed.